File: activemq/openwire-dotnet/src/OpenWire.Client/Connection.cs
using System;
using System.Collections;
using OpenWire.Client.Commands;
using OpenWire.Client.Core;

namespace OpenWire.Client {
        /// <summary>
        /// Represents a connection with a message broker
        /// </summary>
        public class Connection : IConnection {

                private ConnectionInfo info;
                private ITransport transport;
                IList sessions = new ArrayList();
                private bool transacted;
                private bool connected;
                private bool closed;
                private AcknowledgementMode acknowledgementMode;
                private long sessionCounter;

                public Connection(ITransport transport, ConnectionInfo info) {
                        this.transport = transport;
                        this.info = info; 
                }

                /// <summary>
                /// Creates a new session to work on this connection
                /// </summary>
                public ISession CreateSession() {
                        return CreateSession(transacted, acknowledgementMode); 
                }

                /// <summary>
                /// Creates a new session to work on this connection
                /// </summary>
                public ISession CreateSession(bool transacted, AcknowledgementMode acknowledgementMode) {
                        CheckConnected();
                        SessionInfo info = CreateSessionInfo(transacted, acknowledgementMode);
                        SyncRequest(info);
                        Session session = new Session(this, info);
                        sessions.Add(session);
                        return session; 
                }

                public void Dispose() {
                        foreach (Session session in sessions) {
                                session.Dispose(); 
                        }
                        sessions.Clear();
                        transport.Dispose();
                        closed = true; 
                }

                // Properties

                public ITransport ITransport {
                        get { return transport; }
                        set { this.transport = value; } 
                }

                public bool Transacted {
                        get { return transacted; }
                        set { this.transacted = value; } 
                }

                public AcknowledgementMode AcknowledgementMode {
                        get { return acknowledgementMode; }
                        set { this.acknowledgementMode = value; } 
                }

                public string ClientId {
                        get { return info.ClientId; }
                        set {
                                if (connected) {
                                        throw new OpenWireException("You cannot change the ClientId once the Connection is connected"); 
                                }
                                info.ClientId = value;
                        }
                }

                // Implementation methods

                /// <summary>
                /// Performs a synchronous request-response with the broker
                /// </summary>
                public Response SyncRequest(Command command) {
                        CheckConnected();
                        Response response = ITransport.Request(command);
                        if (response is ExceptionResponse) {
                                ExceptionResponse exceptionResponse = (ExceptionResponse) response;
                                // TODO include stack trace
                                throw new OpenWireException("Request failed: " + exceptionResponse); 
                        }
                        return response; 
                }


                protected SessionInfo CreateSessionInfo(bool transacted, AcknowledgementMode acknowledgementMode) {
                        SessionInfo answer = new SessionInfo();
                        SessionId sessionId = new SessionId();
                        sessionId.ConnectionId = info.ConnectionId.Value;
                        lock (this) {
                                sessionId.Value = ++sessionCounter; 
                        }
                        answer.SessionId = sessionId;
                        return answer; 
                }

                protected void CheckConnected() {
                        if (closed) {
                                throw new ConnectionClosedException(); 
                        }
                        if (!connected) {
                                SyncRequest(info);
                                connected = true; 
                        } 
                } 
        } 
}

File: openwire-cpp/src/Connection.cpp
/*
 * Copyright 2006 The Apache Software Foundation or its licensors, as
 * applicable.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include "Connection.hpp"

using namespace apache::activemq::client;


// --- Constructors -------------------------------------------------

/*
 *
 */
Connection::Connection(p<ITransport> transport, p<ConnectionInfo> connectionInfo)
{
    this->transport      = transport ;
    this->connectionInfo = connectionInfo ;
    this->connected      = false ;
    this->transacted     = false ;
    this->closed         = false ;
}

/*
 *
 */
Connection::~Connection()
{
}


// --- Attribute methods --------------------------------------------

/*
 *
 */
p<ITransport> Connection::getTransport()
{
    return transport ;
}

/*
 *
 */
void Connection::setTransport(p<ITransport> transport) throw(OpenWireException) 
{
    this->transport = transport ;
}

/*
 *
 */
p<string> Connection::getClientId()
{
    return connectionInfo->getClientId() ;
}

/*
 *
 */
void Connection::setClientId(const char* value)
{
    if( connected )
        throw OpenWireException("You cannot change the ClientId once a connection is established") ; 

    connectionInfo->setClientId( value ) ;
}

/*
 *
 */
bool Connection::getTransacted()
{
    return transacted ;
}

/*
 *
 */
void Connection::setTransacted(bool tx)
{
    transacted = tx ;
}

/*
 *
 */
AcknowledgementMode Connection::getAcknowledgementMode()
{
    return acknowledgementMode ;
}

/*
 *
 */
void Connection::setAcknowledgementMode(AcknowledgementMode mode)
{
    acknowledgementMode = mode ;
}


// --- Operation methods --------------------------------------------

/*
 *
 */
p<ISession> Connection::createSession() throw(OpenWireException)
{
    return createSession(transacted, acknowledgementMode) ;
}

/*
 *
 */
p<ISession> Connection::createSession(bool transacted, AcknowledgementMode mode) throw(OpenWireException)
{
    checkConnected() ;

    p<SessionInfo> sessionInfo = createSessionInfo(transacted, acknowledgementMode) ;

    // Send session info to broker
    syncRequest(sessionInfo) ;

    p<ISession> session = new Session(this, sessionInfo) ;
    sessions.push_back(session) ;

    return session ;
}

/*
 *
 */
p<Response> Connection::syncRequest(p<ICommand> command) throw(OpenWireException) 
{
    checkConnected() ;
    
    p<Response> response = transport->request(command) ;

    if( response->getCommandType() == ExceptionResponse::TYPE )
    {
        p<ExceptionResponse> exceptionResponse = (p<ExceptionResponse>&)response ;
        p<BrokerError>       brokerError = exceptionResponse->getException() ;
        string               message ;

        // Build error message
        message.assign("Request failed: ") ;
        message.append( brokerError->getExceptionClass()->c_str() ) ;
        message.append(", ") ;
        message.append( brokerError->getStackTrace()->c_str() ) ;

        throw new OpenWireException( message.c_str() ) ; 
    }
    return response ;
}


// --- Implementation methods ---------------------------------------

/*
 *
 */
p<SessionInfo> Connection::createSessionInfo(bool transacted, AcknowledgementMode acknowledgementMode)
{
    p<SessionInfo> sessionInfo = new SessionInfo() ;
    p<SessionId>   sessionId   = new SessionId() ;
    
    sessionId->setConnectionId ( connectionInfo->getConnectionId()->getValue()->c_str() ) ;

    mutex.lock() ;
    sessionId->setValue( ++sessionCounter ) ; 
    mutex.unlock() ;

    sessionInfo->setSessionId( sessionId ) ;
    return sessionInfo ; 
}

/*
 *
 */
void Connection::checkConnected() throw(OpenWireException) 
{
    if( closed )
        throw new ConnectionClosedException("Oops! Connection already closed.") ;

    if( !connected )
    {
        syncRequest((p<ICommand>&)connectionInfo) ;
        connected = true ;
    } 
}
